### PR TITLE
GDS BFF: Auth0 Authentication & Authorization

### PR DESCRIPTION
**Pull request by bbengfort**
### Scope of changes

This story was a bit tougher than I anticipated because the auth0 SDK functionality had to be reconfigured for use with Gin rather than standard http handlers. I created two middlewares:

- `Authenticate`: looks up the access token in the authorization header and parses its claims, validating that the token was created by Auth0. If the token is invalid, returns a 403 error; otherwise it adds the claims to the context. If there is no authorization header then this middleware adds an "anonymous user" claim with no permissions.
- `Authorize`: a middleware that requires specific permissions to be validated; fetches the claims from the context modified by Authenticate and checks the permissions.

I also modified logging to make debugging a bit easier and updated the context for Auth0 configuration.

Note that we could perform a subsequent request to get the userinfo; but instead of doing that, I created rules to add claims directly to the access token (VASP, permissions, and email). This should be sufficient for now, but we can always fetch detailed user information from the backend using secrets if necessary.

Fixes SC-4592

#### Auth0 Debug

I also created a web application for debugging Auth0 access tokens and user profiles via a test SPA (single page application). I've added this app to the repository since I think it will be useful for us as we work with Auth0. This app allows you to login with our TRISA Auth0 tenant (so you can check user creation without running the GDS User UI) and displays an access token and userinfo profile. This was very helpful:

1. Debugging configuration changes on the Auth0 dashboard
2. Grabbing tokens for debugging API authentication

Screen Shot 2022-05-22 at 21 22 23

To use this application run:

```
$ python web/auth0-debug/serve.py
```

Then open [http://localhost:3000](http://localhost:3000)

### Type of change

- [ ] bug fix
- [x] new feature
- [ ] documentation
- [ ] other (describe)

### Acceptance criteria

This is certainly a big PR to check but the questions I'd like the reviewer to address are:

- [x] Are we happy with the test coverage even though we cannot test the happy path for Authenticate?
- [x] Does the separation of concerns between Authenticate and Authorize make sense?
- [x] What format would be best for zerolog messages (replacing "Request" which is the current message)
- [x] What questions do you have for changes made in the code
- [x] Do you understand all the test cases, were any cases missed?

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [x]  I have updated the dependencies list
- [ ]  ~I have recompiled and included new protocol buffers to reflect changes I made~
- [x]  I have added new test fixtures as needed to support added tests
- [x]   Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [x]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -131,18 +131,25 @@ GDS_BFF_LOG_LEVEL=info
...
 GDS_BFF_MAINNET_TIMEOUT=5s
+
 GDS_BFF_DATABASE_URL=trtl://localhost:4436/
 GDS_BFF_DATABASE_REINDEX_ON_BOOT=false
 GDS_BFF_DATABASE_INSECURE=true
 GDS_BFF_DATABASE_CERT_PATH=
 GDS_BFF_DATABASE_POOL_PATH=
-GDS_BFF_TRACK_PERFORMANCE=false
+
+GDS_BFF_SENTRY_TRACK_PERFORMANCE=false
```
Comment (bbengfort) [inline .env.template:152]: Trying to keep the `.env.template` up to date with configuration changes -- organizing it a bit here too.

File: .gitignore [hunk 2]
```diff
@@ -19,3 +19,7 @@
 fixtures/
 .env
 .secrets
+
+# API Testing
+.access_token_cached
+api.py
```
Comment (bbengfort) [inline .gitignore:25]: Some Python scripts I was using to test and debug what I was getting from Auth0 and what the Golang code code expect and parse. I'm happy to provide this script as a private gist on request.

File: go.mod [hunk 3]
```diff
@@ -7,9 +7,9 @@ require (
 	cloud.google.com/go/secretmanager v1.4.0
 	github.com/BurntSushi/toml v1.1.0 // indirect
 	github.com/DataDog/zstd v1.5.0 // indirect
+	github.com/auth0/go-jwt-middleware/v2 v2.0.0
 	github.com/cockroachdb/errors v1.9.0 // indirect
 	github.com/cockroachdb/pebble v0.0.0-20220418160907-785bb36dae6e // indirect
-	github.com/dn365/gin-zerolog v0.0.0-20171227063204-b43714b00db1
```
Comment (bbengfort) [inline go.mod:12]: Nice to get rid of this dependency

File: pkg/bff/auth/auth.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+// AnonymousClaims are used to identify unauthenticated requests that have no permissions.
+var AnonymousClaims = Claims{Scope: ScopeAnonymous, Permissions: nil}
+
+// Claims extracts custom data from the JWT token provided by Auth0
+type Claims struct {
+	Scope       string   `json:"scope"`
+	Permissions []string `json:"permissions"`
+	VASP        string   `json:"https://vaspdirectory.net/vasp"`
+	Email       string   `json:"https://vaspdirectory.net/email"`
+}
```
Comment (bbengfort) [inline pkg/bff/auth/auth.go:36]: Scope comes by default on the access token from Auth0. The Permissions will come if you tick a check mark in the settings of the API. The `VASP` and `Email` claims required me to create Auth0 actions to add them to the access token.

File: pkg/bff/auth/testdata/invalid_token.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImVkOGE0ODcyLWUwNDItNGU2NS04ZDQ3LTFkZDkzYTUwNTI0ZSJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTUxNjIzOTAyMn0.N0aJnJEE8lselYyqIQWupEvw9SBlc2tQ4YFqH6ACaxpnGz2nIbxU4PWP-E-rJ949a1GU24laozNHzyqNpZKrGvzHDeYdS8HcCZLIhgeU7bM1JWpXmMEmmPk-oi9EywVsbJo1Ll9dPcDIGPM_9OTc0MvgKEjLajCH8gA74wRal5QinKSR4VQIJkZ1AVG26SU77Qj8tHynOjoi14KHUghTh9zXc7Vb-mXQo002jTFgbmLl43qOsd7U6BfCTjVAWBoGhKMO3K9jyeeYIWqC8CTOB1eQQMT5F6qsn0t1R392F5tK-KPNmGZAXfvN9D6uVqauUTRgSPqq5rGb4a4DazhQ-w
```
Comment (bbengfort) [inline pkg/bff/auth/testdata/invalid_token.txt:1]: I created this token on jwt.io with a generated RSA key pair - it won't match Auth0's public keys which is why it's invalid.

File: pkg/bff/auth/testdata/validated_claims.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+		"https://vaspdirectory.net/email": "rebecca@example.com"
+	},
+	"RegisteredClaims": {
+		"iss": "https://example.auth0.com/",
+		"sub": "auth0|9d56yPBwjRv1tp3gEIA",
+		"aud": ["https://api.example.net", "https://example.auth0.com/userinfo"],
+		"exp": 1653341676,
+		"iat": 1653255276
+	}
+}
```
Comment (bbengfort) [inline pkg/bff/auth/testdata/validated_claims.json:15]: Also an example of what the expected claims from the Auth0 access token will look like.

File: pkg/bff/config/config.go [hunk 7]
```diff
@@ -131,3 +145,45 @@ func (c SentryConfig) GetRelease() string {
...
+			return nil, errors.New("invalid configuration: could not parse issuer url")
+		}
+	}
+
+	// Default to the HTTPS scheme and reparse domain only configuration.
+	if u, err = url.Parse("https://" + domain); err != nil {
+		return nil, errors.New("invalid configuration: specify auth0 domain of the configured tenant")
+	}
+	return u, nil
+}
```
Comment (bbengfort) [inline pkg/bff/config/config.go:189]: This was kind of dumb - we could have just put a URL in the configuration … but I was led astray by the Auth0 documentation which used `url.Parse` to create a URL from a domain. After I implemented this and realized it wasn't necessary I expanded it to be flexible for the config to be an actual URL, just a domain, and to deal with trailing slashes (hopefully protecting the folks doing devops).

File: pkg/bff/trace.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+// Middleware that tracks request performance with Sentry.
+func (s *Server) TrackPerformance() gin.HandlerFunc {
+	log.Info().Float64("sample_rate", s.conf.Sentry.SampleRate).Msg("sentry performance tracking enabled")
+	return func(c *gin.Context) {
+		span := sentry.StartSpan(c.Request.Context(), c.Request.URL.Path)
+		defer span.Finish()
+		c.Next()
+	}
+}
```
Comment (bbengfort) [inline pkg/bff/trace.go:17]: Moved this so it was a bit easier to find and adapt if necessary.

File: pkg/utils/logger/middleware.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+		switch {
+		case status >= 400 && status < 500:
+			logctx.Warn().Msg(msg)
+		case status > 500:
+			logctx.Error().Msg(msg)
+		default:
+			logctx.Info().Msg(msg)
+		}
+	}
+}
```
Comment (bbengfort) [inline pkg/utils/logger/middleware.go:54]: This was basically all that was in `gin-zerolog` Go package -- but the log message in Google Console was "Request" … you had to dive into the JSON to get any information which was annoying. So now the message contains more info including status code, method, and path. Plus we got rid of a dependency!

File: web/auth0-debug/htdocs/auth_config.json [hunk 10]
```diff
@@ -0,0 +1,5 @@
+{
+  "domain": "dev-bu-hbv3o.us.auth0.com",
+  "clientId": "k3nFXxrQFVinXsGGpdT4dKAtfKYZbQle",
+  "audience": "https://bff.vaspdirectory.net"
+}
```
Comment (bbengfort) [inline web/auth0-debug/htdocs/auth_config.json:5]: This is all public info that would be configured in a single page app (SPA) front-end.

File: web/auth0-debug/htdocs/css/style.css [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    border-top-left-radius: 0;
+    border-top-right-radius: 0;
+}
+
+.gated-content {
+    width: 100%;
+    max-width: 260px;
+    padding: 15px;
+    margin: auto;
+}
```
Comment (bbengfort) [inline web/auth0-debug/htdocs/css/style.css:46]: I wanted the auth0-debug app to be pretty.

File: web/auth0-debug/htdocs/js/app.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    // Process the login state
+    await auth0.handleRedirectCallback();
+
+    updateUI();
+
+    // Use replaceState to redirect the user away and remove querystring parameters
+    window.history.replaceState({}, document.title, "/");
+  }
+
+};
```
Comment (bbengfort) [inline web/auth0-debug/htdocs/js/app.js:74]: Most of this code is from the Auth0 SPA tutorial, only modified slightly.

File: web/auth0-debug/serve.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    with socketserver.TCPServer(("", args.port), handler) as httpd:
+        print(f"web app started at http://localhost:{args.port}")
+        httpd.serve_forever()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="auth0 debug webapp")
+    parser.add_argument("-p", "--port", default=DEFAULT_PORT, type=int)
+    args = parser.parse_args()
+    main(args)
```
Comment (bbengfort) [inline web/auth0-debug/serve.py:24]: A quick and dirty web server for debugging auth0.

File: pkg/bff/server.go [hunk 14]
```diff
@@ -221,38 +217,59 @@ func (s *Server) SetURL(url string) {
...
+		// require a user context; if it doesn't, it should come before authentication.
+		authenticator,
+	}
+
+	// Add the middleware to the router
+	for _, middleware := range middlewares {
+		if middleware != nil {
+			s.router.Use(middleware)
+		}
 	}
```
Comment (bbengfort) [inline pkg/bff/server.go:272]: Above code reorganizes how middleware is added to give us a bit more control of ordering and visibility into the entire middleware chain.

File: web/auth0-debug/serve.py (hunk 13 above)
Comment (pdeziel) [reply web/auth0-debug/serve.py:24]: Hooray for `http.server`

**Review by pdeziel — APPROVED** [review]
I think this looks good, the division between authentication and authorization makes sense to me, as long as we're careful about applying the middlewares in the right order. On that point, I think the middleware refactoring here is a good step in the right direction, but since the ordering is so specific is there a better way to enforce it besides the inline comments?

The logging changes are also great and should give us greater visibility. The SPA debug tool is also very neat and I can see that being helpful.

File: pkg/utils/logger/middleware.go (hunk 9 above)
Comment (bbengfort) [reply pkg/utils/logger/middleware.go:54]: I added the request parameters to the path on line 21 - honestly, this was just copying what gin-zerolog was doing before; do you think we should have the params logged separately from the path?

**Pull request merged**